### ISSUE TITLE
Can't expand headers within `<>`

### ISSUE THREAD

**Issue by ishandutta2007** [Viatorus/quom#52]
I believe the code is designed to expand headers within `""` only.
For example when I am trying to amalgamate this code [primecount api](https://github.com/kimwalisch/primecount/blob/master/test/api.cpp),
since in `#include <primecount.hpp>` the header name `primecount.hpp` is enclosed by `<>` it fails to pull in the required code.
Ideally we should use a lookup table to check if a header is native c++ header or not . Dont you think so?
Please do update if you can get quom to work with the [above ](https://github.com/kimwalisch/primecount)repo.

PS: This ticket has nothing to do with my other [ticket](https://github.com/Viatorus/quom/issues/51). Now I am on macOS and I dont have that issue here.

**Comment by Viatorus** [comment]
Yes, this is currently not possible but your repo would never work reliable. I think you should rework your includes.

For example this one:

https://github.com/kimwalisch/primecount/blob/master/include/gourdon.hpp
```
#include <int128_t.hpp>
#include <print.hpp>

#include <stdint.h>
```

If you use `<...>`, a compiler assumes it is a system header.

By default, g++ for example will first try to find this header file in this order:
- /usr/include/c++/11
-  /usr/include/x86_64-linux-gnu/c++/11
-  /usr/include/c++/11/backward
-  /usr/lib/gcc/x86_64-linux-gnu/11/include
-  /usr/local/include
-  /usr/include/x86_64-linux-gnu
-  /usr/include

After it does not find it inside these folders, it will start the search relative to the file.

P.S. <stdint.h> should be included through <cstdint> (as every c header)

**Comment by ishandutta2007** [comment]
> If it didn't find the header there, it will start the search relative to the file with the #include.

Yes thats what the logic should be, if that logic is already coded why does then `""` or `<>` make a difference ?

> I think you should rework your includes.

We should try to make quom work for as many repos as possible. Changing an entire repo everytime is definitely not a scalable option, if we could upgrade quom with small changes. Among all amalgamate 
 tools I [came across ](https://stackoverflow.com/questions/50887018/cmake-generate-single-header-file) yours seem to be the most general . I am okay to send you a pull request if you dont have bandwidth, but first you should help me understand.

**Comment by Viatorus** [comment]
> why does then "" or <> make a difference ?

- "" starts with a relative search and then searches inside the system includes
- <> starts with a search in the system includes and then a relative searches 

So if you use <> for your relative includes and the system e.g. provides a <min.hpp>, this would be used first.

> We should try to make quom work for as many repos as possible. 

Yes, as long as this project structure is uncommon doesn't have conceptional issues. 

And yours has some issues.

* your include directory contains
  * the public API header primecount.hpp and primecount.h
  * and the private headers

1. Private headers do not belong inside the include folder because typically, the folder will be copied to the system include directory and every header would be public available (same for package manager like conan). 
2. Your include folder requires (because it is common best practice) a subfolder, so a include of your library would be `#include <primecount/primecount.hpp>`

**Comment by ishandutta2007** [comment]
> And yours has some issues.

None of the big repos are mine . If I had written these for my purpose, I would write code in a single file not needing something like quom at all.  As long as it builds sucessfully they won't consider it an issue.  I can't ask all big repos like the above one to change themselves as per quom, quom has to change to support all kinds of repos. And why would they even listen to you if you just call their projet as "conceptional issue"-d when hundreds of people are using those successfully everyday.

**Comment by ishandutta2007** [comment]
> Yes, as long as this project structure is a common one doesn't have conceptional issues.

I tried a couple of other popular repos too. Couldnt get quom to work for any till now. Can you make it woking for this repo ? Say this example https://github.com/flintlib/flint2/blob/trunk/examples/crt.cpp
I couldn't quom here either. For exact issue I may create a separate ticket if you want. But first tell me does this repo fall under the common structure that you are talking about ?

**Comment by ishandutta2007** [comment]
These are the of repos I work with. I am not sure with how many of them quom will work out of the box
https://en.cppreference.com/w/cpp/links/libs
https://en.wikipedia.org/wiki/List_of_numerical_libraries#C++

**Comment by ishandutta2007** [comment]
> does this [FLINTLIB ](https://en.wikipedia.org/wiki/Fast_Library_for_Number_Theory) repo fall under the common structure that you are talking about ?

Did you manage to get it working?

Seems like quom has only been tested on toy projects. To make quom into something of practical utility we should test it against top 10 or top 20 C/C++ open-source libraries. I need amalgamation quite a lot. I do it manually spending days. I have been wanting to start a amalgamation tool for a while. I would definitely want to help you take quom to something of practical utility provided I have your support.

**Comment by ishandutta2007** [comment]
?

**Comment by ishandutta2007** [comment]
Can you kindly reply. It is okay to start a project aimed at toy projects only or keeping community only to self etc.. Different people have different motivation in starting open-source projects. Not all repos have the same goal. Unless you speak up we wont get to know what you intend to do with quom.

**Comment by Viatorus** [comment]
> None of the big repos are mine 

I was only focusing on primecount. I didn't pay attention to any other library. I assumed you are the author of this library primecount.

>  As long as it builds successfully they won't consider it an issue.

Well, it may be no issue if you only build primecount alone for itself but the code structure design has an issue and if any other project does the same and you combine them nothing works as expected anymore.

> And why would they even listen to you if you ... when hundreds of people are using those successfully everyday

A library (like primecount) should not only provide a good algorithm/implementation but also a safe deployment and distribution strategy.

> tested on toy projects.

How dare you to insult anybody's project a toy project?

> But first tell me does this [FLINTLIB ](https://en.wikipedia.org/wiki/Fast_Library_for_Number_Theory)repo fall under the common structure that you are talking about ?

Imho the project structure is still a little mess . Please take a look at libraries like boost, abseil, opencv and poco where I take my orientation from.

> Did you manage to get it working?

I tested it and it fails on several steps. Most are include bugs because many developer do not understand the difference between `""` and `<>`.

1. `flint.h` includes the system include `limits.h` with `""`, which is wrong. Quom cannot find this include in the library repo and fails. At all other places in flint, <limits.h> is used. This one is maybe a typo?! Changed to `<>`.

2. `flint-config.h` isn't found. Okay this would be something CMake would automatic configure. Quom does not interact with CMake. I generate this file manually.

3. Than it fails for finding "gc.h" which is an external library (maybe this one? https://github.com/doublec/gc/blob/master/gc.h - I don't know) . Same as system includes, external libraries must be included with `<>`. Changing to `<>`.

4. Than `fmpz-conversions.h` - again a CMake generated file. I generate this file manually.

Done. 11350 lines of code. And it compiles: https://godbolt.org/z/1ErMqMG56 (if GCC < 11 is used because of a conversion bug).

>  It is okay to start a project aimed at toy projects only or keeping community only to self etc.. Different people have different motivation in starting open-source projects. Not all repos have the same goal.

I think the overall misunderstanding that happend here is that I assumed you are a library author and you assumed Quom is a tool other third party libraries can be unified by just running Quom over it.

Quom is designed for library authors who want to simplify the distribution of there multi-file project. So other can simply download one file and include it in there code base. This need is mainly because of the lack of a widely used package manager for C++. I never tested Quom against any existing library. But there are some libraries out there. One with many stars I now of is: https://github.com/hanickadot/compile-time-regular-expressions

Your need is something different - I assume for performance reason?

I don't think Quom would be ever capable to support libraries like flint out of the box if special configuration/build steps are necessary. Quom follows an include/header/source pattern. If this isn't given or not consistent - Quom will fail. 

There are also other cases where a successfully Quom generated file will never compile. If two source files do have an identical named internal symbol (like `foo.c` and `bar.c` both defines a `static int x = 0;`).

I am open to support more projects - projects that uses `<>` for including their own headers which is not wrong if done right. E.g. like boost is doing it: https://github.com/boostorg/beast/blob/develop/include/boost/beast.hpp#L15 - it could be possible to add options/paths to support this kind of projects.

**Comment by ishandutta2007** [comment]
Thanks for the reply. 

> Imho the project structure is still a little mess . Please take a look at libraries like boost, abseil, opencv and poco where I take my orientation from.

These names that you mentioned are also among top 20-30 opensource libraries as well. Does quom already support these? It would be great if it's tested against these and ensured support for these to start with. If they have been already tested we should call that out in the docs or preferably somewehre in the examples or testing folder.

> How dare you to insult anybody's project a toy project?
it's definitely not about insult, its a common terminology in software industry, we often use the terms "toy test data" or "mock test data" to mean something that has not been tested on real production or production quality data yet.

I assumed you are a library author and you assumed Quom is a tool other third party libraries can be unified by just running Quom over it.
> exactly

Quom is designed for library authors who want to simplify the distribution of there multi-file project.
> I kind of got that from your reply. Running quom on own project doesnt really appeal to me. It may have some small users as you mentioned. My view is as Projects mautre generally the progresseion is from flatter to more higherarchical. So I would always have access to my old flatter project at an time by going back to earlier version. Thats why I dont consider that as strong usecase especially if I have wriiten the full code. I get your idea of simplified distribution vs my requirement of extraction of a subset from multi-level complex project. Do yo know of any other amalgamator tool whose philosophy is more inclined towards mine?
I am still curious what are the benefits of distributing the same code in a flatter structure if want to distrubute the entire thing. it doesnt same space, complicates readability as well.
My usecase is primarily to save source code size.

**Comment by Viatorus** [comment]
> These names that you mentioned are also among top 20-30 open-source libraries as well. Does quom already support these?

Never tested and it is also not the scope of this project. Boost uses `<>` for his own header inclusion (which is fine) but Quom does not yet support this.

> If they have been already tested we should call that out in the docs or preferably somewhere in the examples or tests folder.

Even through Quom could claim at some point, these libraries can be amalgamate minor changes in the sources (duplicated internal symbol) could break all.

> not looking to unify the entire library but only the code that I need.

So not the scope of this project.

> Do yo know of any other amalgamator tool whose philosophy is more inclined towards mine?

No and I never heard of such a use case like yours.

> Now that we both have understanding of each others motivation can you suggest some better approach for my use case.

No, not really. I don't know why you have to save _source code size_. Do you mean binary code size? Can't you link statically or something like that?

**Comment by ishandutta2007** [comment]
> binary code size?

No source code size,  you only have mentioned that the amalgamated file you generated was 11K lines while the original repo size was 890K lines.

> No and I never heard of such a use case like yours. 

And I have never heard of such use case that you are trying to sell. No wonder the repo is struggling to build a community around it.

**Comment by ishandutta2007** [comment]
In fact `gcc -E` almost does the job for me, only that I don't need to print the codes from system headers.
So My idea of building a amalgamator is just to tweak `gcc -E` such that it suppresses the codes from system headers.
As long as `gcc` compiles successfully so would `gcc -E` and so would `gcc -E -my_new_option`.
In such case we need not worry about structure, format, tokenizer, parenthesis or <>, interaction type with CMake or any other BS whatsoever.

**Comment by ishandutta2007** [comment]
> your approch too doesn't pull in the entire 890K lines either but only the functons that are being called from the initiating program. So your claim that you are amalgamating the entire library is not true.

I thought you would at least respond to this since it does more like what I expected an amalgamator to do that what you intended to build it for.

Anyways nevermind lets come back to the issue under the ticket, I have reworked all the imports [here](https://github.com/ishandutta2007/primecount) (check my fork only) . Let me know if now satisfies the project structure you desire.

**Comment by Viatorus** [comment]
> I thought you would at least respond to this since it does more like what I expected an amalgamator to do that what you intended to build it for.
> So your claim that you are amalgamating the entire library is not true.

This wan't the intention behind this test. If someone wants to amalgamating awhole library, all public headers must be included.

> Anyways nevermind lets come back to the issue under the ticket, I have reworked all the imports [here](https://github.com/ishandutta2007/primecount) (check my fork only) . Let me know if it now satisfies the project structure you desire.

> Anyways nevermind lets come back to the issue under the ticket, I have reworked all the imports [here](https://github.com/ishandutta2007/primecount) (check my fork only) . Let me know if it now satisfies the project structure you desire.

Changing <> to "" looks okay - does it work for you now with Quom? 

But why aren't you just using the provided build instruction? Sorry, I still don't get your use case why you want/must do this. Maybe its my poor English or my limited experience in your field of work. I don't know.
